raiden-ts/src/path/epics.ts

```diff
@@ -1,48 +1,198 @@
+      error_details: t.type({
+        from_: Address,
+        to: Address,
+        value: UInt(32),
+      }),
+    }),
+  ]),
+);
+interface PathError extends t.TypeOf<typeof PathError> {}
+
+class InvalidPFSRequestError extends Error {
+  readonly iou: Signed<IOU> | undefined;
+  readonly errorCode: number;
+  constructor(message: string, errorCode: number, iou: Signed<IOU> | undefined) {
+    super(message);
+    this.iou = iou;
+    this.errorCode = errorCode;
+  }
+}
+
+const makeIOU = (
+  sender: Address,
+  receiver: Address,
+  amount: UInt<32>,
+  chainId: UInt<32>,
+  oneToNAddress: Address,
+  blockNumber: number,
+): IOU => ({
+  sender: sender,
+  receiver: receiver,
+  chain_id: chainId,
+  amount: amount,
+  one_to_n_address: oneToNAddress,
+  expiration_block: bigNumberify(blockNumber).add(2 * 10 ** 5) as UInt<32>,
+});
+
+const updateIOU = (iou: IOU, price: UInt<32>): IOU => ({
+  ...iou,
+  amount: iou.amount.add(price) as UInt<32>,
+});
+
+const signIOU$ = (iou: IOU, signer: Signer): Observable<Signed<IOU>> =>
+  from(signer.signMessage(
+    concat([
+      encode(iou.one_to_n_address, 20),
+      encode(iou.chain_id, 32),
+      encode(MessageTypeId.IOU, 32),
+      encode(iou.sender, 20),
+      encode(iou.receiver, 20),
+      encode(iou.amount, 32),
+      encode(iou.expiration_block, 32),
+    ]),
+  ) as Promise<Signature>).pipe(map(signature => ({ ...iou, signature })));
+
+const makeAndSignLastIOURequest$ = (sender: Address, receiver: Address, signer: Signer) =>
+  defer(() => {
+    const payload = {
+      sender,
+      receiver,
+      timestamp: new Date(Date.now()).toISOString().split('.')[0],
+    };
+
+    const timestamp = hexlify(toUtf8Bytes(payload.timestamp));
+    const messageHash = concat([
+      encode(payload.sender, 20),
+      encode(payload.receiver, 20),
+      encode(timestamp, hexDataLength(timestamp)),
+    ]);
+    return from(signer.signMessage(messageHash) as Promise<Signature>).pipe(
+      map(signature => ({
+        ...payload,
+        signature,
+      })),
+    );
+  });
+
+const prepareNextIOU$ = (
+  state: RaidenState,
+  pfs: PFS,
+  deps: RaidenEpicDeps,
+  httpTimeout: number,
+  tokenNetwork: Address,
+): Observable<Signed<IOU>> => {
+  const cachedIOU: IOU | undefined = get(state.path, ['iou', tokenNetwork, pfs.address]);
+  return (cachedIOU
+    ? of(cachedIOU)
+    : makeAndSignLastIOURequest$(state.address, pfs.address, deps.signer).pipe(
+        mergeMap(payload =>
+          fromFetch(
+            `${pfs.url}/api/v1/${tokenNetwork}/payment/iou?${new URLSearchParams(
+              payload,
+            ).toString()}`,
+            {
+              method: 'GET',
+              headers: { 'Content-Type': 'application/json' },
+            },
+          ),
+        ),
+        timeout(httpTimeout),
+        mergeMap(async response => {
+          const text = await response.text();
+          const oneToNAddress = memoize(
```

Comment:
This needs to be defined outside function, or else `memoize` is getting re-created every time and will just memory-leak instead of helping cache the async call

Reply:
So moving it to top-level should be enough right